custom/build.gradle.kts

```diff
@@ -6,14 +6,15 @@ dependencies {
   implementation(project(":common"))
   implementation(project(":inferred-spans"))
   implementation(project(":universal-profiling-integration"))
-  compileOnly(project(":bootstrap"))
   implementation(project(":resources"))
   compileOnly("io.opentelemetry:opentelemetry-sdk")
   compileOnly("io.opentelemetry:opentelemetry-sdk-extension-autoconfigure-spi")
   compileOnly("io.opentelemetry.javaagent:opentelemetry-javaagent-extension-api")
   compileOnly("io.opentelemetry.javaagent:opentelemetry-javaagent-tooling")
   compileOnly(libs.bundles.semconv)
 
+  implementation(libs.contribSpanStacktrace)
```

Comment:
Excluding transitive dependencies is close to use `compileOnly` instead of `implementation`: we compile against the API, but don't pull transitive dependencies when used.

The downside of this approach is that we then have to pull back needed dependencies again, for example with tests in this module require an extra `testImplementation`, and all the modules that depend on `custom` will have to do the same.

In a sense we need to have the ability to express "use this for implementation, but some transitive dependencies need to be filtered-out in packaging", and I don't know if we can do that with a custom dependency type.

Reply:
> The downside of this approach is that we then have to pull back needed dependencies again, for example with tests in this module require an extra testImplementation, and all the modules that depend on custom will have to do the same.

My understanding is that `custom` is not intended to be used by anyone but as an aggregator for building our agent-distro and the extension. both of those don't want to have the transitive dependencies.

For having to pull in the transitive dependencies for tests:
Simply adding `testImplementation(libsContribSpanStacktrace)` should pull in everything that has been excluded.
Alternatively, the commonly used OpenTelemetry modules for tests could be added to the test classpath via our `testing-common` module or an convention.

The main downside I see with manually removing the "bad" classes at packaging time is that it makes the build process depend on package names, which are neither public nor guaranteed to be stable (e.g. `io/opentelemetry/extension/incubator/`), making the build potentially prone to errors which are hard to find.

In my opinion the extension (e.g. span-stacktrace) should not be pulling in SDK dependencies via `implementation` anyway and use `compileOnly` , because it doesn't make just our live harder: It also makes the life harder for people who want to package those extensions as their own agent-extension fatJar, becuase they have to do the same exclusions.
For people who use the SDK directly I'd expect them to add the SDK to their classpath explicitly and not rely on transitive dependencies of an SDK-extension, which is kind of the wrong way around.

Reply:
This might also be something to discuss with the Otel Java SIG to see if they have any opinion on it. Looking at the contrib repo, most extensions there pull in the SDK + API as `API`, which has the downsides (and IMO no upsides) explained above.